# Sales Data Exploratory Analysis


## 1. Data Understanding



In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
rawsales = pd.read_csv('salesdata.csv')

I´ve successfully loaded the csv into the dataframe variable (rawsales)

In [3]:
print(rawsales)

     Order ID  Amount  Profit  Quantity         Category      Sub-Category  \
0     B-26776    9726    1275         5      Electronics  Electronic Games   
1     B-26776    9726    1275         5      Electronics  Electronic Games   
2     B-26776    9726    1275         5      Electronics  Electronic Games   
3     B-26776    4975    1330        14      Electronics          Printers   
4     B-26776    4975    1330        14      Electronics          Printers   
...       ...     ...     ...       ...              ...               ...   
1189  B-26370    8825    3594        15        Furniture            Tables   
1190  B-26298    2082     642         8      Electronics            Phones   
1191  B-26298    2082     642         8      Electronics            Phones   
1192  B-26298    2082     642         8      Electronics            Phones   
1193  B-25068     914     163        13  Office Supplies           Markers   

     PaymentMode  Order Date        CustomerName     State     

In [4]:
rawsales.head()

,Order ID,Amount,Profit,Quantity,Category,Sub-Category,PaymentMode,Order Date,CustomerName,State,City,Year-Month
0,B-26776,9726,1275,5,Electronics,Electronic Games,UPI,2023-06-27,David Padilla,Florida,Miami,2023-06
1,B-26776,9726,1275,5,Electronics,Electronic Games,UPI,2024-12-27,Connor Morgan,Illinois,Chicago,2024-12
2,B-26776,9726,1275,5,Electronics,Electronic Games,UPI,2021-07-25,Robert Stone,New York,Buffalo,2021-07
3,B-26776,4975,1330,14,Electronics,Printers,UPI,2023-06-27,David Padilla,Florida,Miami,2023-06
4,B-26776,4975,1330,14,Electronics,Printers,UPI,2024-12-27,Connor Morgan,Illinois,Chicago,2024-12


In [5]:
rawsales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1194 entries, 0 to 1193
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Order ID      1194 non-null   object
 1   Amount        1194 non-null   int64 
 2   Profit        1194 non-null   int64 
 3   Quantity      1194 non-null   int64 
 4   Category      1194 non-null   object
 5   Sub-Category  1194 non-null   object
 6   PaymentMode   1194 non-null   object
 7   Order Date    1194 non-null   object
 8   CustomerName  1194 non-null   object
 9   State         1194 non-null   object
 10  City          1194 non-null   object
 11  Year-Month    1194 non-null   object
dtypes: int64(3), object(9)
memory usage: 112.1+ KB


The dataset contains 1,194 observations across 12 variables. There are no immediately apparent null values. Three variables are numeric, while nine are stored as objects. Order Date is currently represented as an object rather than a datetime variable and may require conversion before time-series analysis.

In [6]:
rawsales.shape

(1194, 12)

Rows = 1194 
Columns = 12

In [7]:
rawsales.head(10)

,Order ID,Amount,Profit,Quantity,Category,Sub-Category,PaymentMode,Order Date,CustomerName,State,City,Year-Month
0,B-26776,9726,1275,5,Electronics,Electronic Games,UPI,2023-06-27,David Padilla,Florida,Miami,2023-06
1,B-26776,9726,1275,5,Electronics,Electronic Games,UPI,2024-12-27,Connor Morgan,Illinois,Chicago,2024-12
2,B-26776,9726,1275,5,Electronics,Electronic Games,UPI,2021-07-25,Robert Stone,New York,Buffalo,2021-07
3,B-26776,4975,1330,14,Electronics,Printers,UPI,2023-06-27,David Padilla,Florida,Miami,2023-06
4,B-26776,4975,1330,14,Electronics,Printers,UPI,2024-12-27,Connor Morgan,Illinois,Chicago,2024-12
5,B-26776,4975,1330,14,Electronics,Printers,UPI,2021-07-25,Robert Stone,New York,Buffalo,2021-07
6,B-26942,1525,185,12,Office Supplies,Pens,Debit Card,2024-05-11,John Fields,Florida,Orlando,2024-05
7,B-26942,1525,185,12,Office Supplies,Pens,Debit Card,2021-10-09,Clayton Smith,Florida,Miami,2021-10
8,B-26640,883,117,10,Electronics,Laptops,EMI,2022-11-18,Richard Kelley,California,Los Angeles,2022-11
9,B-26640,8127,3551,16,Furniture,Tables,Credit Card,2022-11-18,Richard Kelley,California,Los Angeles,2022-11


This shows the first 10 entries in the dataset

In [8]:
rawsales.describe()

,Amount,Profit,Quantity
count,1194.000000,1194.000000,1194.000000
mean,5178.089615,1348.992462,10.674204
std,2804.921955,1117.992573,5.777102
min,508.000000,50.000000,1.000000
25%,2799.000000,410.000000,6.000000
50%,5152.000000,1014.000000,11.000000
75%,7626.000000,2035.000000,16.000000
max,9992.000000,4930.000000,20.000000


- The dataset contains 1194 observations across all 3 numerical variables, .info() shows that there are no missing values recognized by pandas in these column.

- describe() has made profits distribution worth investigating

- The middle 50% of Amount values fall between 2,799 and 7,626 as shown by the 25th and 75th percentiles.

- Profit has a wide and and considerable range from 50 to 4930.

In [9]:
rawsales.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1189    False
1190    False
1191    False
1192    False
1193    False
Length: 1194, dtype: bool

In [10]:
rawsales.duplicated().sum()

np.int64(0)

No exact duplicate rows were identified in the dataset

In [11]:
rawsales.isnull()

,Order ID,Amount,Profit,Quantity,Category,Sub-Category,PaymentMode,Order Date,CustomerName,State,City,Year-Month
0,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...
1189,False,False,False,False,False,False,False,False,False,False,False,False
1190,False,False,False,False,False,False,False,False,False,False,False,False
1191,False,False,False,False,False,False,False,False,False,False,False,False
1192,False,False,False,False,False,False,False,False,False,False,False,False


In [12]:
rawsales.isnull().sum()

Order ID        0
Amount          0
Profit          0
Quantity        0
Category        0
Sub-Category    0
PaymentMode     0
Order Date      0
CustomerName    0
State           0
City            0
Year-Month      0
dtype: int64

0 null values across all 12 columns were detected

In [13]:
rawsales["Category"].unique()

array(['Electronics', 'Office Supplies', 'Furniture'], dtype=object)

This shows what the unique values are in the column.

In [14]:
rawsales["Category"].nunique()

3

This tells us how many unique values there are in the column.

In [15]:
rawsales["Category"].value_counts()

Category
Furniture          407
Office Supplies    399
Electronics        388
Name: count, dtype: int64

In [16]:
rawsales["PaymentMode"].unique()

array(['UPI', 'Debit Card', 'EMI', 'Credit Card', 'COD'], dtype=object)

In [17]:
rawsales["PaymentMode"].value_counts()

PaymentMode
Debit Card     260
Credit Card    258
UPI            252
EMI            218
COD            206
Name: count, dtype: int64

There are 5 payments methods.

Debit card was the most frequent with 260 records.

In [18]:
rawsales["State"].unique()

array(['Florida', 'Illinois', 'New York', 'California', 'Texas', 'Ohio'],
      dtype=object)

In [19]:
rawsales["State"].value_counts()

State
New York      226
California    218
Florida       200
Texas         189
Illinois      181
Ohio          180
Name: count, dtype: int64

In [20]:

rawsales["PaymentMode"].groupby(rawsales["State"]).value_counts()

State       PaymentMode
California  Credit Card    48
            UPI            47
            Debit Card     45
            EMI            41
            COD            37
Florida     UPI            46
            Credit Card    43
            COD            42
            Debit Card     39
            EMI            30
Illinois    UPI            45
            EMI            39
            Credit Card    38
            Debit Card     31
            COD            28
New York    Debit Card     57
            UPI            53
            Credit Card    47
            COD            35
            EMI            34
Ohio        Debit Card     42
            Credit Card    39
            EMI            36
            UPI            33
            COD            30
Texas       Debit Card     46
            Credit Card    43
            EMI            38
            COD            34
            UPI            28
Name: count, dtype: int64

In [21]:
pd.crosstab(rawsales["State"], rawsales["PaymentMode"])

PaymentMode,COD,Credit Card,Debit Card,EMI,UPI
State,,,,,
California,37,48,45,41,47
Florida,42,43,39,30,46
Illinois,28,38,31,39,45
New York,35,47,57,34,53
Ohio,30,39,42,36,33
Texas,34,43,46,38,28


In [22]:
pd.crosstab(rawsales["State"], rawsales["PaymentMode"], margins=True, normalize="index")

PaymentMode,COD,Credit Card,Debit Card,EMI,UPI
State,,,,,
California,0.169725,0.220183,0.206422,0.188073,0.215596
Florida,0.210000,0.215000,0.195000,0.150000,0.230000
Illinois,0.154696,0.209945,0.171271,0.215470,0.248619
New York,0.154867,0.207965,0.252212,0.150442,0.234513
Ohio,0.166667,0.216667,0.233333,0.200000,0.183333
Texas,0.179894,0.227513,0.243386,0.201058,0.148148
All,0.172529,0.216080,0.217755,0.182580,0.211055


In [23]:
pd.crosstab(rawsales["State"], rawsales["PaymentMode"], margins=True, normalize="index") * 100

PaymentMode,COD,Credit Card,Debit Card,EMI,UPI
State,,,,,
California,16.972477,22.018349,20.642202,18.807339,21.559633
Florida,21.000000,21.500000,19.500000,15.000000,23.000000
Illinois,15.469613,20.994475,17.127072,21.546961,24.861878
New York,15.486726,20.796460,25.221239,15.044248,23.451327
Ohio,16.666667,21.666667,23.333333,20.000000,18.333333
Texas,17.989418,22.751323,24.338624,20.105820,14.814815
All,17.252931,21.608040,21.775544,18.257956,21.105528


In [24]:
pd.crosstab(rawsales["State"], rawsales["PaymentMode"], margins=True, normalize="index").mul(100).round(2)

PaymentMode,COD,Credit Card,Debit Card,EMI,UPI
State,,,,,
California,16.97,22.02,20.64,18.81,21.56
Florida,21.00,21.50,19.50,15.00,23.00
Illinois,15.47,20.99,17.13,21.55,24.86
New York,15.49,20.80,25.22,15.04,23.45
Ohio,16.67,21.67,23.33,20.00,18.33
Texas,17.99,22.75,24.34,20.11,14.81
All,17.25,21.61,21.78,18.26,21.11


Rounding up to 2 decimal point


Payment-method distribution varies somewhat between states. Florida has the highest proportion of COD transactions at 21%, while debit-card transactions are most prominent in New York (25.22%) and Texas (24.34%). UPI shows a notable difference between states, ranging from 14.81% in Texas to 24.86% in Illinois. In contrast, credit-card usage is comparatively consistent across all six states, remaining around 21–23%

In [25]:
print(rawsales["Order Date"])

0       2023-06-27
1       2024-12-27
2       2021-07-25
3       2023-06-27
4       2024-12-27
           ...    
1189    2024-07-31
1190    2020-06-02
1191    2022-12-15
1192    2020-08-07
1193    2024-10-26
Name: Order Date, Length: 1194, dtype: object


In [26]:
rawsales.rename(columns={"order date": "Order_Date"}, inplace=True)

In [27]:
rawsales.columns

Index(['Order ID', 'Amount', 'Profit', 'Quantity', 'Category', 'Sub-Category',
       'PaymentMode', 'Order Date', 'CustomerName', 'State', 'City',
       'Year-Month'],
      dtype='object')

In [28]:
rawsales = rawsales.rename(columns={"Order ID": "Order_ID", "Order Date": "Order_Date"})

In [29]:
rawsales.columns

Index(['Order_ID', 'Amount', 'Profit', 'Quantity', 'Category', 'Sub-Category',
       'PaymentMode', 'Order_Date', 'CustomerName', 'State', 'City',
       'Year-Month'],
      dtype='object')

In [30]:
rawsales["Order_Date"].head(10)

0    2023-06-27
1    2024-12-27
2    2021-07-25
3    2023-06-27
4    2024-12-27
5    2021-07-25
6    2024-05-11
7    2021-10-09
8    2022-11-18
9    2022-11-18
Name: Order_Date, dtype: object

In [31]:
rawsales["Order_Date"] = pd.to_datetime(rawsales["Order_Date"])

In [32]:
rawsales["Order_Date"].dtype

dtype('<M8[ns]')

In [33]:
rawsales["Order_Date"].dt.to_period("M").astype(str)

0       2023-06
1       2024-12
2       2021-07
3       2023-06
4       2024-12
         ...   
1189    2024-07
1190    2020-06
1191    2022-12
1192    2020-08
1193    2024-10
Name: Order_Date, Length: 1194, dtype: object

In [34]:
YM_check = rawsales["Order_Date"].dt.to_period("M").astype(str) == rawsales["Year-Month"]

In [35]:
YM_check.value_counts()

True    1194
Name: count, dtype: int64

In [36]:
rawsales.columns

Index(['Order_ID', 'Amount', 'Profit', 'Quantity', 'Category', 'Sub-Category',
       'PaymentMode', 'Order_Date', 'CustomerName', 'State', 'City',
       'Year-Month'],
      dtype='object')

In [37]:
rawsales["Order_ID"].nunique()

547

In [38]:
rawsales["Order_ID"].value_counts()

Order_ID
B-26489    12
B-26032    12
B-25465    12
B-26279    10
B-25715     9
           ..
B-26545     1
B-26603     1
B-25945     1
B-26565     1
B-25068     1
Name: count, Length: 547, dtype: int64

In [39]:
rawsales[rawsales["Order_ID"] == "B-26489"]

,Order_ID,Amount,Profit,Quantity,Category,Sub-Category,PaymentMode,Order_Date,CustomerName,State,City,Year-Month
354,B-26489,2791,894,14,Electronics,Laptops,COD,2022-10-08,Brett Sutton,Texas,Austin,2022-10
355,B-26489,2791,894,14,Electronics,Laptops,COD,2023-10-27,Anna Blackburn,Ohio,Cincinnati,2023-10
356,B-26489,2791,894,14,Electronics,Laptops,COD,2021-02-09,Michelle Bailey,Texas,Dallas,2021-02
357,B-26489,6223,1478,3,Furniture,Sofas,Debit Card,2022-10-08,Brett Sutton,Texas,Austin,2022-10
358,B-26489,6223,1478,3,Furniture,Sofas,Debit Card,2023-10-27,Anna Blackburn,Ohio,Cincinnati,2023-10
359,B-26489,6223,1478,3,Furniture,Sofas,Debit Card,2021-02-09,Michelle Bailey,Texas,Dallas,2021-02
360,B-26489,8943,3121,17,Furniture,Chairs,Debit Card,2022-10-08,Brett Sutton,Texas,Austin,2022-10
361,B-26489,8943,3121,17,Furniture,Chairs,Debit Card,2023-10-27,Anna Blackburn,Ohio,Cincinnati,2023-10
362,B-26489,8943,3121,17,Furniture,Chairs,Debit Card,2021-02-09,Michelle Bailey,Texas,Dallas,2021-02
363,B-26489,3297,102,12,Electronics,Phones,Credit Card,2022-10-08,Brett Sutton,Texas,Austin,2022-10


Order_ID is not a unique transaction identifier because identical IDs occur across difference customers, dates and locations. Repeated Order_ID values can contain identical product-level Amount, Profit, Quantity and PaymentMode combinations across otherwise distinct customer transactions.nd locations.

In [40]:
rawsales.columns

Index(['Order_ID', 'Amount', 'Profit', 'Quantity', 'Category', 'Sub-Category',
       'PaymentMode', 'Order_Date', 'CustomerName', 'State', 'City',
       'Year-Month'],
      dtype='object')

With quality checks and cleaning performed to an agreed percentage. The next phase of this analysis would seek to answer some business questions. 

Overall Performance - How much revenue and profit does the dataset represent? How many units were sold? What does profitability look like?

Product Performance - Which categories/sub-categories generate the most revenue, profit and quantity?

Geographic Performance - Which states/cities contribute most to revenue and profit?

Time Performance - How do revenue and profit change over time? Are there strong/weak months or years?

Relationships - Does higher Amount generally correspond to higher Profit? Does Quantity have much relationship with either?

Before any grouping occurs, I'll like to answer; What are the total amount?, total profit? and total quantity? represented by all 1194 records.

In [41]:
total_amount = rawsales["Amount"].sum()

total_profit = rawsales["Profit"].sum()

total_quantity = rawsales["Quantity"].sum()

In [42]:
print("Total Amount:", total_amount)
print("Total Profit:", total_profit)
print("Total Quantity:", total_quantity)

Total Amount: 6182639
Total Profit: 1610697
Total Quantity: 12745


Total amount = 6,182,639.

Total profit = 1,610,697.

Total quantity = 12,745.

Bear in mind 1,194 rows ≠ 1,194 orders

The dataset records a total Amount of approximately 6.18 million, total Profit of approximately 1.61 million, and a total Quantity of 12,745 units across 1,194 records.

In [43]:
# Calculate the profit margin percentage using the formula:
profit_margin = (total_profit / total_amount) * 100

what proportion of the total amount is represented by profit, then convert that proportion to a percentage by multiplying by 100. 
This gives you the profit margin percentage, which indicates how much profit is generated for every dollar of sales.

In [44]:
print(round(profit_margin, 2))

26.05


Print the profit margin percentage rounded to two decimal places.

So, if Amount represents the sales base against which Profit is measured, the dataset has an overall profit-to-amount ratio of approximately 26.05%.

Total Amount by Category

In [61]:
total_amount_by_category = rawsales.groupby("Category")["Amount"].sum()

In [62]:
print(total_amount_by_category)

Category
Electronics        2054456
Furniture          2038673
Office Supplies    2089510
Name: Amount, dtype: int64


Amount is distributed relatively evenly across the three product categories. Office Supplies records the highest total Amount at approximately 2.09 million, while Furniture records the lowest at approximately 2.04 million. Why does Furniture appear in more records but generates a lower total Amount, what comparisons can be made to understand why?. Below we shall explore the average rather than the sum total above.

Average Amount by Category

In [63]:
average_amount_by_category = rawsales.groupby("Category")["Amount"].mean()

In [64]:
print(average_amount_by_category)

Category
Electronics        5294.989691
Furniture          5009.024570
Office Supplies    5236.867168
Name: Amount, dtype: float64


Furniture records have a lower average Amount (5,009) than Electronics (5,295) and Office Supplies (5,237). we already discovered that a row doesn't necessarily equal one identifiable sale/order given the weirdness with Order_ID.

Total and average Amounts by Sub-category

In [47]:
total_subcat = rawsales.groupby("Sub-Category")["Amount"].sum().sort_values(ascending=False)

average_subcat = rawsales.groupby("Sub-Category")["Amount"].mean().sort_values(ascending=False)

In [48]:
print("Total Amount by Sub-Category: ", [total_subcat])

Total Amount by Sub-Category:  [Sub-Category
Markers             627875
Tables              625177
Sofas               568367
Printers            566359
Electronic Games    565092
Pens                552269
Paper               524755
Phones              503055
Chairs              431964
Laptops             419950
Bookcases           413165
Binders             384611
Name: Amount, dtype: int64]


In [49]:
print("Average Amount by Sub-Category:", [average_subcat])

Average Amount by Sub-Category: [Sub-Category
Printers            5961.673684
Markers             5707.954545
Electronic Games    5433.576923
Binders             5417.056338
Chairs              5204.385542
Tables              5124.401639
Paper               5045.721154
Sofas               4985.675439
Phones              4980.742574
Pens                4844.464912
Laptops             4772.159091
Bookcases           4695.056818
Name: Amount, dtype: float64]


Markers generated the highest total Amount (627,875), closely followed by Tables (625,177). However, Printers recorded the highest average Amount per record (5,961.67), despite ranking fourth by total Amount. This suggests that total performance is influenced not only by the value of individual records but also by how frequently each sub-category occurs

Are Markers and Tables generating high total Amount partly because they appear more frequently in the dataset?

In [50]:
rawsales["Sub-Category"].unique()

array(['Electronic Games', 'Printers', 'Pens', 'Laptops', 'Tables',
       'Chairs', 'Markers', 'Sofas', 'Paper', 'Binders', 'Phones',
       'Bookcases'], dtype=object)

In [51]:
Record_Count = rawsales["Sub-Category"].value_counts()

In [67]:
subcat_comparison = pd.concat(
    [Record_Count, total_subcat, average_subcat], axis=1
)

In [68]:
subcat_comparison.columns = ["Record_Count", "Total_Amount", "Average_Amount"]

In [69]:
subcat_comparison

,Record_Count,Total_Amount,Average_Amount
Sub-Category,,,
Tables,122,625177,5124.401639
Pens,114,552269,4844.464912
Sofas,114,568367,4985.675439
Markers,110,627875,5707.954545
Electronic Games,104,565092,5433.576923
Paper,104,524755,5045.721154
Phones,101,503055,4980.742574
Printers,95,566359,5961.673684
Laptops,88,419950,4772.159091


Sub-categories with more records generally tend to produce higher Total Amounts. However, record frequency alone does not determine total performance; Average Amount also plays an important role. Markers combine a relatively high record count with a high Average Amount, resulting in the highest Total Amount, while Printers demonstrate that a high Average Amount can compensate for a lower record count.

In [55]:
subcat_comparison["Average_Amount"] = subcat_comparison["Average_Amount"].round(2)

In [56]:
subcat_comparison

,Record_Count,Total_Amount,Average_Amount
Sub-Category,,,
Tables,122,625177,5124.40
Pens,114,552269,4844.46
Sofas,114,568367,4985.68
Markers,110,627875,5707.95
Electronic Games,104,565092,5433.58
Paper,104,524755,5045.72
Phones,101,503055,4980.74
Printers,95,566359,5961.67
Laptops,88,419950,4772.16


Analysis update

The dataset contains 1,194 records covering product transactions across three main categories: Electronics, Furniture, and Office Supplies.

Initial data-quality checks found no missing values and no exact duplicate rows. The Order_Date column was converted from an object to a datetime datatype. The existing Year-Month column was verified against the year-month derived from Order_Date, with all 1,194 records matching, making the column redundant for the analysis.

Investigation of Order_ID found 547 unique IDs across 1,194 records. Individual IDs can occur across different customers, dates and locations, meaning Order_ID should not be treated as a reliable unique identifier for individual transactions without further documentation.

For overall performance, the dataset records a Total Amount of 6,182,639, Total Profit of 1,610,697, and Total Quantity of 12,745. Profit represents approximately 26.05% of the total Amount, assuming these fields are intended to be compared on that basis.

At category level, total Amount is relatively evenly distributed:

Office Supplies    2,089,510

Electronics        2,054,456

Furniture          2,038,673

Despite Furniture having the highest number of records (407), it has the lowest total Amount. Its average Amount per record (5,009.02) is also lower than Electronics (5,294.99) and Office Supplies (5,236.87), helping explain this difference.

At sub-category level, Markers generated the highest Total Amount (627,875), narrowly followed by Tables (625,177). However, Printers had the highest Average Amount per record (5,961.67) despite ranking fourth in Total Amount.

Comparing record frequency with average and total Amount suggests that both volume and value per record contribute to overall sub-category performance. Tables benefit from having the highest record count (122), while Printers demonstrate that a high Average Amount can compensate for fewer records (95). Markers combine a relatively high record count (110) with the second-highest Average Amount (5,707.95), resulting in the highest Total Amount.

What categories actually generate the most profit?, this is what we aim to answer next.

In [70]:
total_profit_by_category = rawsales.groupby("Category")["Profit"].sum().sort_values(ascending=False)

total_amount_by_category = rawsales.groupby("Category")["Amount"].sum().sort_values(ascending=False)

In [71]:
print(total_profit_by_category)

print(total_amount_by_category)

Category
Office Supplies    551575
Furniture          540542
Electronics        518580
Name: Profit, dtype: int64
Category
Office Supplies    2089510
Electronics        2054456
Furniture          2038673
Name: Amount, dtype: int64


In [72]:
category_comparison = pd.concat(
    [total_profit_by_category, total_amount_by_category], axis=1
)

In [73]:
category_comparison.columns = ["Total_Profit", "Total_Amount"]

In [74]:
category_comparison

,Total_Profit,Total_Amount
Category,,
Office Supplies,551575,2089510
Furniture,540542,2038673
Electronics,518580,2054456


Office Supplies generated the highest Total Profit at 551,575, consistent with it also having the highest Total Amount. However, Furniture generated more Profit than Electronics despite recording a lower Total Amount. This suggests that higher Amount does not necessarily correspond to higher profitability across categories.

In [78]:
category_comparison["profit_margin"] = (category_comparison["Total_Profit"] / category_comparison["Total_Amount"]) * 100

In [91]:
rounded_category_comparison = round(category_comparison, 2)

print(rounded_category_comparison)

                 Total_Profit  Total_Amount  profit_margin
Category                                                  
Office Supplies        551575       2089510          26.40
Furniture              540542       2038673          26.51
Electronics            518580       2054456          25.24


Furniture has the highest profit margin at 26.51%, despite having the lowest Total Amount of the three categories. Office Supplies follows closely with a 26.40% margin while generating both the highest Total Amount and Total Profit. Electronics ranks lowest in profitability, with a 25.24% profit margin despite generating a higher Total Amount than Furniture.

In [82]:
total_profit_by_subcat = rawsales.groupby("Sub-Category")["Profit"].sum().sort_values(ascending=False)


print(total_profit_by_subcat)
print(total_subcat)

Sub-Category
Markers             174749
Tables              156796
Paper               149723
Electronic Games    148454
Printers            146259
Sofas               142854
Pens                129846
Chairs              122892
Bookcases           118000
Phones              113607
Laptops             110260
Binders              97257
Name: Profit, dtype: int64
Sub-Category
Markers             627875
Tables              625177
Sofas               568367
Printers            566359
Electronic Games    565092
Pens                552269
Paper               524755
Phones              503055
Chairs              431964
Laptops             419950
Bookcases           413165
Binders             384611
Name: Amount, dtype: int64


In [83]:
total_subcategory_comparison = pd.concat(
    [total_profit_by_subcat, total_subcat], axis=1
)

In [84]:
total_subcategory_comparison.columns = ["Total Profit by Subcategory", "Total Amount by Subcategory"]

In [85]:
total_subcategory_comparison

,Total Profit by Subcategory,Total Amount by Subcategory
Sub-Category,,
Markers,174749,627875
Tables,156796,625177
Paper,149723,524755
Electronic Games,148454,565092
Printers,146259,566359
Sofas,142854,568367
Pens,129846,552269
Chairs,122892,431964
Bookcases,118000,413165


In [87]:
total_subcategory_comparison["subcat_profit_margin"] = (total_subcategory_comparison["Total Profit by Subcategory"] / total_subcategory_comparison["Total Amount by Subcategory"]) * 100

In [93]:
rounded_subcat_comparison = round(total_subcategory_comparison, 2)

print(rounded_subcat_comparison)

                  Total Profit by Subcategory  Total Amount by Subcategory  \
Sub-Category                                                                 
Markers                                174749                       627875   
Tables                                 156796                       625177   
Paper                                  149723                       524755   
Electronic Games                       148454                       565092   
Printers                               146259                       566359   
Sofas                                  142854                       568367   
Pens                                   129846                       552269   
Chairs                                 122892                       431964   
Bookcases                              118000                       413165   
Phones                                 113607                       503055   
Laptops                                110260                   

Profit Performance

Office Supplies generated the highest Total Profit at 551,575, followed by Furniture at 540,542 and Electronics at 518,580. However, Furniture recorded the highest category-level profit margin at 26.51%, narrowly ahead of Office Supplies at 26.40%, while Electronics had the lowest at 25.24%.

At sub-category level, Markers generated both the highest Total Amount (627,875) and Total Profit (174,749), while maintaining a relatively strong profit margin of 27.83%.

Profit margin reveals a different picture from total performance. Bookcases had the highest profit margin at 28.56%, closely followed by Paper (28.53%) and Chairs (28.45%). In contrast, Phones had the lowest margin at 22.58%.

These results show that the sub-categories generating the greatest total Profit are not necessarily those with the highest profit margins. Total Profit reflects overall contribution, while profit margin shows Profit relative to Amount.

Which categories account for the greatest quantity of products in the dataset?

In [96]:
total_quantity_by_category = rawsales.groupby("Category")["Quantity"].sum().sort_values(ascending=False)

print(total_quantity_by_category)

Category
Furniture          4441
Electronics        4258
Office Supplies    4046
Name: Quantity, dtype: int64


Furniture's performance appears more volume-driven, whereas Office Supplies generates a greater Amount despite a lower overall Quantity.

In [98]:
average_quantity_by_category = rawsales.groupby("Category")["Quantity"].mean().sort_values(ascending=False)

print(average_quantity_by_category)

Category
Electronics        10.974227
Furniture          10.911548
Office Supplies    10.140351
Name: Quantity, dtype: float64


In [99]:
quantity_comparison = pd.concat(
    [total_quantity_by_category, average_quantity_by_category], axis=1
)

In [100]:
quantity_comparison.columns = ["Total_Quantity", "Average_Quantity"]

In [101]:
quantity_comparison

,Total_Quantity,Average_Quantity
Category,,
Furniture,4441,10.911548
Electronics,4258,10.974227
Office Supplies,4046,10.140351


In [112]:
total_quantity_by_subcategory = rawsales.groupby("Sub-Category")["Quantity"].sum().sort_values(ascending=False)

average_quantity_by_subcategory = rawsales.groupby("Sub-Category")["Quantity"].mean().sort_values(ascending=False)

rounded_average_quantity_by_subcategory = average_quantity_by_subcategory.round(2)

In [113]:
quantity_by_subcat_comparison = pd.concat(
    [total_quantity_by_subcategory, rounded_average_quantity_by_subcategory], axis=1
)

In [114]:
quantity_by_subcat_comparison.columns = ["Total_Quantity_by_Subcategory", "Average_Quantity_by_Subcategory"]

In [115]:
quantity_by_subcat_comparison

,Total_Quantity_by_Subcategory,Average_Quantity_by_Subcategory
Sub-Category,,
Tables,1303,10.68
Sofas,1233,10.82
Electronic Games,1220,11.73
Pens,1204,10.56
Markers,1173,10.66
Printers,1124,11.83
Bookcases,1030,11.70
Paper,981,9.43
Phones,980,9.70


Markers aren't winning because they have the highest quantity. Their strength seems to come from a combination of relatively high record frequency, high average Amount and a strong profit margin.

Quantity Performance

Furniture recorded the highest Total Quantity at 4,441 units, followed by Electronics with 4,258 and Office Supplies with 4,046. However, Electronics had the highest Average Quantity per record at 10.97, narrowly ahead of Furniture at 10.91, while Office Supplies had the lowest average at 10.14.

This suggests that Furniture's higher Total Quantity is partly supported by its greater number of records, rather than substantially higher quantities per record. Office Supplies, meanwhile, recorded both the lowest Total Quantity and Average Quantity despite generating the highest Total Amount and Total Profit, indicating that its overall performance is not primarily driven by product volume.

At sub-category level, Tables recorded the highest Total Quantity at 1,303 units, followed by Sofas (1,233) and Electronic Games (1,220). However, Printers had the highest Average Quantity per record at 11.83, followed by Electronic Games (11.73) and Bookcases (11.70).

The difference between total and average quantity again highlights the effect of record frequency. Tables have the highest Total Quantity while averaging 10.68 units per record, whereas Printers have a lower Total Quantity of 1,124 but the highest Average Quantity. Therefore, high total product volume can result from either a greater number of records, larger quantities per record, or a combination of both.